src/stdlib_str2num.f90

```diff
@@ -0,0 +1,532 @@
+        ! Verify leading negative
+        sign = 1
+        if( iachar(s(p:p)) == minus_sign+digit_0 ) then
+            sign = -1 ; p = p + 1
+        end if
+        if( iachar(s(p:p)) == Inf ) then
+            v = sign*ieee_value(v,  ieee_positive_inf); return
+        else if( iachar(s(p:p)) == NaN ) then
+            v = ieee_value(v,  ieee_quiet_nan); return
+        end if
+        !----------------------------------------------
+        ! read whole and fractional number in a single integer
+        pP = 127
+        int_wp = 0
+        do i = p, min(maxdpt+p-1,len(s))
+            val = iachar(s(i:i))-digit_0
+            if( val >= 0 .and. val <= 9 ) then
+                int_wp = int_wp*10 + val
+            else if( val == period ) then
+                pP = i
+            else
+                exit
+            end if
+        end do
+        pE = i ! Fix the exponent position
+        do while( i<=len(s) )
+           val = iachar(s(i:i))-digit_0
+           if( val < 0 .or. val > 9 ) exit
+           i = i + 1
+        end do
+        p = i
+        resp = pE-min(pP,p) ! If no decimal indicator found it is taken as being in the current p position
+        if( resp <= 0 ) resp = resp+1 
+        !----------------------------------------------
+        ! Get exponential
+        sige = 1
+        if( p<len(s) ) then
+            if( any([le,BE,ld,BD]+digit_0==iachar(s(p:p))) ) p = p + 1
+            if( iachar(s(p:p)) == minus_sign+digit_0 ) then
+                sige = -1
+                p = p + 1
+            else if( iachar(s(p:p)) == plus_sign+digit_0 ) then
+                p = p + 1
+            end if
+        end if
+        
+        i_exp = 0
+        do while( p<=len(s) )
+            val = iachar(s(p:p))-digit_0
+            if( val >= 0 .and. val <= 9) then
+                i_exp = i_exp*10_ikind + val ; p = p + 1
+            else
+                exit
+            end if
+        end do
+        
+        exp_aux = nwnb-1+resp-sige*i_exp
+        if( exp_aux>0 .and. exp_aux<=nwnb+nfnb) then
+            v = sign*int_wp*expbase(exp_aux)
+        else
+            v = sign*int_wp*10._dp**(sige*i_exp-resp+1)
+        end if
+        stat = 0
+    end subroutine
+
+    elemental subroutine to_real_dp(s,v,p,stat)
+        integer, parameter :: wp    = dp
+        !> Sequentially unroll the character and get the sub integers composing the whole number, fraction and exponent
+        ! -- In/out Variables
+        character(*), intent(in) :: s !> input string
+        real(wp), intent(inout)  :: v !> Output real value
+        integer(1), intent(out)  :: p !> last position within the string
+        integer(1), intent(out)  :: stat !> status upon success or failure to read
+
+        ! -- Internal Variables
+        integer(kind=ikind), parameter :: nwnb = 40 !> number of whole number factors
+        integer(kind=ikind), parameter :: nfnb = 64 !> number of fractional number factors
+        integer :: e
```

Comment:
For all these internal implementations, I always prefer to be explicit. For example, if the user compiles the library with `int64` as default (instead of `int32`, would the implementation still work?

Reply:
I see, you are right! For these parameters what I have seen is that it should be a number not too big in order to not saturate the stack. The values I suggested for the number of powers of ten to keep in the stack were found empirically, in any case I think that having ikind = int8 should work as the values are smaller than 127.